### PR TITLE
DBZ-2363 - Add support for pg_logical_emit messages in Postgres connector for the pgoutput plugin

### PR DESCRIPTION
**Pull request by lairen**
https://issues.redhat.com/browse/DBZ-2363

This PR introduces support for `pg_logical_emit_message` messages in the Postgres connector for the pgouput plugin through a configuration property (`logical_decoding_mesage.handling.mode` - defaults to `skip`) Support for these messages has been added to the pgoutput plugin in [Postgres version 14](https://www.postgresql.org/docs/14/protocol-logicalrep-message-formats.html) (https://github.com/postgres/postgres/commit/ac4645c0157fc5fcef0af8ff571512aa284a2cec). 

From [PG14 Documentation](https://www.postgresql.org/docs/14/functions-admin.html)
Screen Shot 2021-11-15 at 2 30 48 PM

Since these are generic messages inserted into the WAL, they are not associated with a table or schema. Although the codepaths for this new message type are completely separate from the other messages, the overall schema had to be updated so that the `schemaName` and `tableName` are nullable.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Thanks @lairen! I've pushed one commit to resolve the formatting issue (please always run `mvn process-sources` before pushing).

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot @lairen! Overall this looks great, a few comments/questions inline. My main concern is around the change to the `source` struct schema, which I don't think we can do. Other than that, the documentation should be updated as well (the new connector option, and the new event type in [this section](https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-events)). Thanks again!

File: debezium-core/src/main/java/io/debezium/data/Envelope.java [hunk 1]
```diff
@@ -50,7 +50,11 @@
         /**
          * An operation that resulted in an existing table being truncated in the source.
          */
-        TRUNCATE("t");
+        TRUNCATE("t"),
+        /**
+         * An operation that resulted in a generic message posted directly to the replication slot
+         */
+        LOGICAL_DECODING_MESSAGE("m");
```
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-core/src/main/java/io/debezium/data/Envelope.java:57]: Great point. I'll update that 👍

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java [hunk 2]
```diff
@@ -135,6 +135,11 @@ public void updateWalPosition(Lsn lsn, Lsn lastCompletelyProcessedLsn, Instant c
         sourceInfo.update(lsn, commitTime, txId, tableId, xmin);
     }
 
+    public void updateWalPosition(Lsn lsn, Lsn lastCompletelyProcessedLsn, Instant commitTime, Long txId, Long xmin) {
```
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java:138]: This was needed because the `updateWalPosition` function defined above takes the `tableId` argument and passes it to `sourceInfo.update()` which does not respect the null value for the `tableId`. In tests, we were seeing these messages having a defined tableId. 

https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/SourceInfo.java#L114-L128

This is the new `update()` function that was added to `SourceInfo` to explicitly set the `tableName` and `schemaName` to `null`

https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/SourceInfo.java#L150-L158

That was the original thought process, but this may no longer be needed if we set the `schemaName` and `tableName` to `N/A` for these messages.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSourceInfoStructMaker.java [hunk 3]
```diff
@@ -19,8 +19,8 @@ public PostgresSourceInfoStructMaker(String connector, String version, CommonCon
         super(connector, version, connectorConfig);
         schema = commonSchemaBuilder()
                 .name("io.debezium.connector.postgresql.Source")
-                .field(SourceInfo.SCHEMA_NAME_KEY, Schema.STRING_SCHEMA)
-                .field(SourceInfo.TABLE_NAME_KEY, Schema.STRING_SCHEMA)
+                .field(SourceInfo.SCHEMA_NAME_KEY, Schema.OPTIONAL_STRING_SCHEMA)
+                .field(SourceInfo.TABLE_NAME_KEY, Schema.OPTIONAL_STRING_SCHEMA)
```
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSourceInfoStructMaker.java:23]: Just ran some tests and it seems like we can set the Schema and Table strings to "N/A" for these new messages, so this schema change will not be necessary 👍

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java [hunk 4]
```diff
@@ -232,6 +232,34 @@ else if (message.getOperation() == Operation.COMMIT) {
...
+                            commitMessage(partition, offsetContext, lsn);
+                        }
+                        return;
+                    }
+
+                    // non-transactional messages will not have a commitTime or txId
+                    if (message.isLastEventForLsn()) {
+                        offsetContext.updateWalPosition(lsn, lastCompletelyProcessedLsn, null, null,
+                                taskContext.getSlotXmin(connection));
+                        commitMessage(partition, offsetContext, lsn);
```
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:248]: In the case of a non-transactional message, there aren't any BEGIN or COMMIT messages surrounding them, so it seemed appropriate to use the same logic that's used in the COMMIT block above.
https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java#L229-L233

The seems to just update the offsetContext and dispatch a heartbeat message, but I can remove that call to `commitMessage` if my assumptions are off base here. 
https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java#L351-L356

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresOffsetContext.java:138]: So in order to set the `tableName` and `schemaName` to an empty string or "N/A" without using these overloaded functions, I passed 
`TableId tableId = PostgresSchema.parse("N/A.N/A")` and passed that to this `updateWalPosition` function here https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java#L246

While that technically works and allows us to get rid of these overloaded functions, it feels a bit "hacky". Another option is to keep the overloaded functions then set the `tableName` and `schemaName` directly in the `SourceInfo` class. These new overloaded functions could also be used for other potential message types that may not be associated with a table.  Any thoughts on those two different approaches?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java [hunk 5]
```diff
@@ -553,6 +567,48 @@ private void decodeTruncate(ByteBuffer buffer, TypeRegistry typeRegistry, Replic
...
+        // Byten The content of the logical decoding message.
+
+        boolean isTransactional = buffer.get() == 1;
+        final Lsn lsn = Lsn.valueOf(buffer.getLong()); // LSN
+        String prefix = readString(buffer);
+        int contentLength = buffer.getInt();
+        byte[] content = new byte[contentLength];
+        for (int i = 0; i < contentLength; i++) {
+            content[i] = buffer.get();
+        }
```
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java]: This solution did work, so I've updated the branch.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java [hunk 6]
```diff
@@ -2881,6 +2885,83 @@ public void testStreamingWithNumericReplicationSlotName() throws Exception {
...
+        start(PostgresConnector.class, configBuilder.build());
+        assertConnectorIsRunning();
+
+        assertRecordsFromSnapshot(2, 1, 1);
+
+        // emit non transactional logical decoding message
+        TestHelper.execute("SELECT pg_logical_emit_message(false, 'foo', 'bar');");
+
+        SourceRecords records = consumeRecordsByTopic(1);
+        List<SourceRecord> recordsForTopic = records.recordsForTopic(topicName("logical_decoding_message"));
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:2925]: How is the topic name determined actually? This should contain the "logical server name" of the connector (the "dbserver1" prefix in many of the tests/examples). Perhaps `<logical name>.message` would make sense (i.e. omitting the implementation detail of this being based on logical decoding)?

Comment (lairen) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:2925]: I like that. The topic suffix is set in the `LogicalDecodingMessageMonitor`  similar to the `TransactionMonitor`. I can update that to be `message`. https://github.com/debezium/debezium/blob/89eb71032167b15906e2a78fed14f9500a56e982/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/LogicalDecodingMessageMonitor.java#L48-L56

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot for the update, @lairen! Overall, this looks great. A few more follow-up questions/suggestions inline, with the most relevant open questions being:

* Should there be the binary "include messages" option vs. a filter-based approach
* Handling of transactional messages may be off in one corner case

I've also pushed a commit for some small adjustments I did along the way. Thanks again, I think we're almost there.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java (hunk 6 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java]: Cool, thanks!

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 7]
```diff
@@ -1176,6 +1178,118 @@ there are no ordering guarantees for the change events pertaining to a table (_c
...
+            "db": "postgres",
+            "schema": "",
+            "table": "",
+            "lsn": 46523128,
+            "xmin": null
+        },
+        "op": "m", // <2>
+        "ts_ms": 1559033904961 // <3>
+        "message": { // <4>
+            "transactional": false,
```
[the remark this answers is not in the document]
Comment (lairen) [reply documentation/modules/ROOT/pages/connectors/postgresql.adoc:1244]: That's fair. My initial thought was to put as much information as possible, but I think as long as the absence of `source.txId` is documented then we don't need the extra boolean and can save the space 👍

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:248]: In the case of a transactional logical message, `isLastEventForLsn` would be `false` since the assumption is that a transactional logical decoding message will eventually be followed by a `COMMIT` message which would trigger the `commitMessage` call in the above if statement. We should only enter this branch if the logical decoding message is not transactional. 

https://github.com/debezium/debezium/blob/8007cc3b053d8299ccbd68d8dbde0a259dc52900/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/LogicalDecodingMessage.java#L35-L38

Your suggestion of checking `message.getTransactionId()` does bring up something I haven't mentioned with this new` LogicalDecodingMessage` class though. I was not able to set the `transactionId` of this new message type to `null` in the case of non-transactional messages because the type is `long` instead of `Long`, so `getTransactionId()` can never be null. In the case of a non-transactional message, the `transactionId` is just the ID of the last `BEGIN` message that the decoder encountered. We're obfuscating all of this by just setting the `txID` to an empty string in the `SourceInfo`. Do you think we can get away with changing the type of `getTransactionId()` in the `ReplicationMessage` interface to `Long` in a minor version update? If so, I can make that update in this PR as well. That feels more correct to me.

https://github.com/debezium/debezium/blob/8007cc3b053d8299ccbd68d8dbde0a259dc52900/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java#L232-L235

https://github.com/debezium/debezium/blob/8007cc3b053d8299ccbd68d8dbde0a259dc52900/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java#L601-L607

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 8]
```diff
@@ -893,6 +965,17 @@ public static AutoCreateMode parse(String value, String defaultValue) {
                     "When 'snapshot.mode' is set as custom, this setting must be set to specify a fully qualified class name to load (via the default class loader)."
                             + "This class must implement the 'Snapshotter' interface and is called on each app boot to determine whether to do a snapshot and how to build queries.");
 
+    public static final Field LOGICAL_DECODING_MESSAGE_HANDLING_MODE = Field.create("logical_decoding_message.handling.mode")
```
[the remark this answers is not in the document]
Comment (lairen) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:968]: I like that a lot, especially since it mimics the behavior for the `table.<include|exclude>.list`. That prior art should make it easy to implement in this PR. I'll put that together soon 👍

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:248]: > In the case of a transactional logical message, isLastEventForLsn would be false since the assumption is [...]

Ah yes, got it now. Probably worth a clarifying comment along those lines on that if statement.

> Do you think we can get away with changing the type of getTransactionId() in the ReplicationMessage interface to Long in a minor version update?

Yes, that's no problem at all; all these  types are not meant for external consumption/implementation, so we take the liberty to modify them as we see fit (admittedly, there is no precise separation between public API/SPI of Debezium and internal parts (yet), but `ReplicationMessage` definitely is internal. I'd only suggest to change this to `OptionalLong` rather than `Long`, so to make make its semantics more explicit.

**Comment by gunnarmorling** [comment]
Hey @lairen, I think I've replied to all your questions; a quick heads-up: we are planning to do the 1.8.0.Beta1 release Thursday this week. So if we have this wrapped up by then, we could include it into that release. If not (and that's no problem at all, just want to share the timeline), the next one would be CR1 on Dec 9th. Whatever works best for you.

**Comment by lairen** [comment]
Hi @gunnarmorling I've just pushed up a commit that implements a prefix filtering approach for the logical decoding messages along with returning `OptionalLong` for `getTransactionId()` that you mentioned earlier. If everything looks good to you, I'd love to be able to get this feature added in this week's 1.8.0.Beta1 release.

**Comment by gunnarmorling** [comment]
Hey @lairen, sorry for the noise from CI, pushing some commits to this PR atm (will squash them later on).

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java [hunk 9]
```diff
@@ -224,14 +224,31 @@ private void processMessages(ChangeEventSourceContext context, PostgresPartition
                     offsetContext.updateWalPosition(lsn, lastCompletelyProcessedLsn, message.getCommitTime(), message.getTransactionId(), null,
                             taskContext.getSlotXmin(connection));
                     if (message.getOperation() == Operation.BEGIN) {
-                        dispatcher.dispatchTransactionStartedEvent(partition, Long.toString(message.getTransactionId()), offsetContext);
+                        dispatcher.dispatchTransactionStartedEvent(partition, message.getTransactionId().toString(), offsetContext);
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:227]: This is invoking `toString()` on `OptionalLong` now, which also causes that test failure of `shouldRegularlyFlushLsnWithTxMonitoring()`. I'll take care of it and also push some further smaller changes.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java [hunk 10]
```diff
@@ -218,7 +226,7 @@ public ChainedLogicalStreamBuilder optionsWithoutMetadata(ChainedLogicalStreamBu
     private void handleBeginMessage(ByteBuffer buffer, ReplicationMessageProcessor processor) throws SQLException, InterruptedException {
         final Lsn lsn = Lsn.valueOf(buffer.getLong()); // LSN
         this.commitTimestamp = PG_EPOCH.plus(buffer.getLong(), ChronoUnit.MICROS);
-        this.transactionId = buffer.getInt();
+        this.transactionId = (long) buffer.getInt();
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java:229]: @jpechane, wondering whether we actually did have an issue here; IIRC, tx ids are unsigned int32. We may have exported values larger than Integer.MAX_VALUE as negative values before?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/LogicalDecodingMessageIT.java [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            assertArrayEquals("txn_bar".getBytes(),
+                    message.getBytes(LogicalDecodingMessageMonitor.DEBEZIUM_LOGICAL_DECODING_MESSAGE_CONTENT_KEY));
+        });
+    }
+
+    @Test
+    @FixFor("DBZ-2363")
+    @SkipWhenDecoderPluginNameIsNot(value = SkipWhenDecoderPluginNameIsNot.DecoderPluginName.PGOUTPUT, reason = "Only supported on PgOutput")
+    @SkipWhenDatabaseVersion(check = LESS_THAN, major = 14, minor = 0, reason = "Message not supported for PG version < 14")
+    public void shouldApplyBinaryHandlingMode() throws Exception {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/LogicalDecodingMessageIT.java:178]: Added this test for making sure the binary handling mode option is applied.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java (hunk 10 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java]: Ok, pushed a commit for converting to unsigned long.

**Comment by gunnarmorling** [comment]
@lairen, I've pushed a few more commits, I *think* this should be good to go now. In terms of the release, we had to push 1.8.Beta1 to early next week. Unless any more issues show up, I'll make sure to include this one. Thanks a lot for this very nice feature addition!

**Comment by gunnarmorling** [comment]
Closing this in favour https://github.com/debezium/debezium/pull/2980, where I have squashed some commits and rebased everything to the latest main branch. There may still be some issue in CI, the last few builds got stuck in some test. I'll investigate, if needed.

Thanks a lot, @lairen! I'll ping you over on the other PR if there's more to discuss.
